src/api/web3.ts

```diff
@@ -0,0 +1,210 @@
+import { Block } from 'web3-eth'
+import { provider } from 'web3-providers'
+import { TransactionReceipt, Transaction } from 'web3-core'
+
+import { windowLoaded } from './utils'
+
+declare global {
+    interface Window {
+        ethereum: provider & {
+            enable: () => Promise<string[]>
+            on?: () => EventListener
+        }
+        web3: Web3
+    }
+}
+
+export interface Web3FrontendAPI {
+    web3: Web3
+    getCurrentAccount: () => Promise<string>
+    getAccounts: () => Promise<string[]>
+    getBalance: (account?: string) => Promise<string>
+    getBlock: (blockNumberOrString: string | number) => Promise<Partial<Block>>
+    getNetwork: () => Promise<string | number>
+    getNetworkName: (id?: number | string) => Promise<string>
+    getNetworkProviderURL: (id: number | string) => string
+    getTransaction: (txHash: string) => Promise<Transaction>
+    getTransactionReceipt: (txHash: string) => Promise<TransactionReceipt>
+    // Utils
+    isAddress: (address: string) => boolean
+    isReadOnly: () => boolean
+    // Provider Specific
+    resetProvider(): void
+    setProvider(provider: string): void
+}
+
+export interface Web3SetupOptions {
+    customProvider?: provider | string
+    reloadOnAccountsChange?: boolean
+}
+
+let web3API: Web3FrontendAPI
+
+let readOnly: boolean
+
+export const getWeb3API = async (options: Web3SetupOptions = {}, force?: boolean): Promise<Web3FrontendAPI> => {
+    if (!force && web3API) return web3API
+
+    return setupWeb3(options)
+}
+
+const getProvider = async ({ customProvider, reloadOnAccountsChange }: Web3SetupOptions): Promise<Web3> => {
+    let provider: Web3
+
+    // wait window load then check window globals...
+    await windowLoaded
+
+    // Use Custom Provider if passed in
+    if (customProvider) provider = new Web3(customProvider)
+    // No Custom Provider? Check window.ethereum
+    else if (window && window.ethereum) {
+        provider = new Web3(window.ethereum)
+
+        // enable Provider, if possible
+        await window.ethereum.enable().catch((error: Error) => {
+            throw new Error(error.message)
+        })
+
+        if (window.ethereum.on && reloadOnAccountsChange) {
+            window.ethereum.on('accountsChanged', function() {
+                window.location.reload()
+            })
+        }
+    }
+    // Check for window.web3's current provider (if it exists)
+    else if (window.web3 && window.web3.currentProvider) provider = new Web3(window.web3.currentProvider)
+    // Else setup localhost version
+    else {
+        try {
+            const url = 'http://localhost:8545'
+            await fetch(url)
+            console.debug('API/WEB3.ts => local node active')
+            provider = new Web3(url)
+        } catch (error) {
+            if (error.readyState === 4 && (error.status === 400 || error.status === 200)) {
+                // the endpoint is active
+                console.debug('API/WEB3.ts catch block => Success')
```

Comment:
What does this mean?
Will be `provider = new Web3(url)` always initialized in this case?

Reply:
for `readOnly` mode, yes

Reply:
I don't get:
1. What is a null provider `provider = new Web3(null)`

2. The logic: if the fetch fails, what is this branch doing. It seems complicated:
![image](https://user-images.githubusercontent.com/2352112/63780522-891f2b80-c8e8-11e9-82e4-d3b340ffb76c.png)

3. This logic. Are we always using metamask if it's installed or do we allow the user to choose. For example, if we want to use wallet connect, it will prompt in the metamask extension.
Also, do we really need to hardcode the development ganache as a fallback?

![image](https://user-images.githubusercontent.com/2352112/63780832-05197380-c8e9-11e9-89bd-a04d6b88d1d5.png)



3.